services/queue/src/data.js

```diff
@@ -189,320 +190,430 @@ class Task {
-    description: Entity.types.Text,
-    stability: Entity.types.String,
-  },
-  migrate(item) {
-    item.lastDateActive = new Date(2000, 0, 1);
-
-    return item;
-  },
-}).configure({
-  version: 4,
-  properties: {
-    provisionerId: Entity.types.String,
-    // the time at which this provisioner should no longer be displayed
-    expires: Entity.types.Date,
-    lastDateActive: Entity.types.Date,
-    description: Entity.types.Text,
-    stability: Entity.types.String,
-    actions: Entity.types.JSON,
-  },
-  migrate(item) {
-    item.actions = [];
-
-    return item;
-  },
-});
-
-/**
- * Expire Provisioner entries.
- *
- * Returns a promise that all expired Provisioner entries have been deleted
- */
-Provisioner.expire = async function(now) {
-  assert(now instanceof Date, 'now must be given as option');
-  let count = 0;
-
-  await Entity.scan.call(this, {
-    expires: Entity.op.lessThan(now),
-  }, {
-    limit: 250, // max number of concurrent delete operations
-    handler: entry => { count++; return entry.remove(true); },
-  });
-
-  return count;
-};
-
-/** Return JSON representation of provisioner meta-data */
-Provisioner.prototype.json = function() {
-  return {
-    provisionerId: this.provisionerId,
-    expires: this.expires.toJSON(),
-    lastDateActive: this.lastDateActive.toJSON(),
-    description: this.description,
-    stability: this.stability,
-    actions: this.actions,
-  };
-};
+class Provisioner {
+  // (private constructor)
+  constructor(props) {
+    Object.assign(this, props);
+  }
+
+  // Create a single instance from a DB row
+  static fromDb(row) {
+    return new Provisioner({
+      provisionerId: row.provisioner_id,
+      expires: row.expires,
+      lastDateActive: row.last_date_active,
+      description: row.description,
+      stability: row.stability,
+      actions: row.actions,
+    });
+  }
+
+  // Create a single instance, or undefined, from a set of rows containing zero
+  // or one elements.  This matches the semantics of get_provisioner
+  static fromDbRows(rows) {
+    if (rows.length === 1) {
+      return Provisioner.fromDb(rows[0]);
+    }
+  }
+
+  // Create an instance from createProvisioner API request arguments
+  static fromApi(provisioner, input) {
+    assert(input.expires);
+    assert(input.lastDateActive);
+    return new Provisioner({
+      provisioner,
+      ...input,
+      // convert to dates
+      expires: new Date(input.expires),
+      lastDateActive: new Date(input.lastDateActive),
+      actions: input.actions || [],
+    });
+  }
+
+  // Get a provisioner from the DB, or undefined
+  static async get(db, provisionerId) {
+    return Provisioner.fromDbRows(await db.fns.get_queue_provisioner(provisionerId));
+  }
+
+  // Call db.get_queue_provisioners.
+  // The response will be of the form { rows, continationToken }.
+  // If there are no provisioners to show, the response will have the
+  // `rows` field set to an empty array.
+  static async getProvisioners(
```

Comment:
It would be nice to replace this (and similar things for the other two classes) with `paginatedIterator` and `paginateResults` as in #3279.  But that's fine to leave as a followup.